### ISSUE TITLE
URL Being cached as IP Address

### ISSUE THREAD

**Issue by Noodleyman** [cubecart/v6#535]
With SSL enabled on a large site. If bots crawl the site using IP address rather than domain name, Internal url's get cached as IP addresses.

Shouldn't be a problem, however when using SSL this causes a certificate error because the IP address does not match the domain name.

In theory pages should only be cached if the URL used to access the page matches the store URL as defined in the store configuration

**Comment by Noodleyman** [comment]
work around is to use htaccess to redirect to the main site, however I think this condition should be handled out of the box, rather than needing to apply a work around via htaccess

  #### REDIRECT IP ADDRESS ####
  RewriteCond %{HTTP_HOST} ^111\.222\.111\.111
  RewriteRule (.*) https://www.myWebSite.co.uk/$1 [R=301,L]

**Comment by bhsmither** [comment]
In my experience up to CC6, and minimally with CC6, it's the "Shop by Category" that has the STORE_URL included in the content of the cached file:
`cache_prefix.html.skin_name.menu.lang_code.cache`
or `8c302.html.foundation.menu.en-US.cache`

If the STORE_URL can be removed and make the links such that it becomes the browser's responsibility to fill in the missing pieces, I think that would solve this problem.

**Comment by Noodleyman** [comment]
Hi Brian,

It was primarily noted in the store with categories as you have noted, however I did also see the problem with links to products.

On small sites I don't think it's a problem, but on large sites it can becomes more notable as the issue gets scaled up. The site this issue cropped up on this week has 40,000 products and several hundred categories.

If the link was changed to allow the browser to fill in the missing pieces, I don't think the CubeCart feature to force SSL would work if visiting a non secure page as the URL gets modified from http to https, but I think it would definitely solve this issue.

There are several ways to approach this as I see.

1) Remove the domain name from the URL's and let the browser fill it in, but this might impact http -> https, would require testing.

2) A htaccess rule to rewrite IP addresses to domain names, but this may impact people who wish to use an IP address to access their store.

3) Only write pages to cache if the domain in the URL used to access the pages/store matches the domain from store settings.

4) Use bot detection code when looking to write to cache. If the user is detected as a bot, do not cache the page. (This might work well when combined with #2)

There are probably other/better solutions, the ones above are just off the top of my head.

**Comment by abrookbanks** [comment]
Hmm... this is tricky. No solution is perfect. I hate the prospect of losing portability and by default most of the URL's seem to be absolute.

**Comment by abrookbanks** [comment]
I'm sorry but I just can't think of a viable or solid solution to this. I would suggest that if it does happen then code along these lines are added to the .htaccess file.

```
RewriteEngine on
RewriteCond %{HTTP_HOST} !^www.example.com$
RewriteRule ^(.*)$ http://www.example.com/$1 [R=301]
```

**Comment by Noodleyman** [comment]
Why would the following option not be viable?

"Only write pages to cache if the domain in the URL used to access the pages/store matches the domain from store settings."

the htaccess rule you added only redirects non www to www.

the correct work around for the ip caching issue is:

  #### REDIRECT IP ADDRESS ####
  RewriteCond %{HTTP_HOST} ^111\.111\.111\.111
  RewriteRule (.*) https://www.example.co.uk/$1 [R=301,L]

Cheers,

**Comment by abrookbanks** [comment]
The redirect rule I suggested redirects anything **other than www.example.com** to www.example.com this includes IP addresses.

**Comment by abrookbanks** [comment]
.. this means the website can only be accessed with your specified domain.

**Comment by bhsmither** [comment]
Another possible fix is to use a different function (or change the function) when creating the `NAVIGATION_TREE`.

In `GUI->_displayNavigation()`, a call is made to `GUI->_maketree()`.

`GUI->_maketree()` and `GUI->_displayNavigation()` both make use of `SEO->buildURL()`.

`SEO->buildURL()` hardcodes `$GLOBALS['storeURL']` into the return string.

`GUI->_displayNavigation()` then caches the `$content` of the template `box.navigation.php` to the file `'html.'.$this->_skin.'.menu.'.$GLOBALS['language']->current()`.

Switching off the cache under specific conditions is a solution, but I feel not the most optimal.

I think not using `SEO->buildURL()` and instead using SEO->generatePath() could be the better solution. (Needs testing.) (Needs to switch the order of arguments between buildURL() and generatePath() and add a couple more.)

**Comment by bhsmither** [comment]
Changing `SEO->buildURL()` by removing `$GLOBALS['storeURL']` from the return string would need testing.

**Comment by abrookbanks** [comment]
I think thats a better solution but as you say it would need testing and I'm in fear of introducing bugs. We can't afford v6 to get a reputation for being buggy. We would need a series of A B testing but even that could miss something due to various server configs.

**Comment by Noodleyman** [comment]
@abrookbanks , this issue appears to still exist in the latest CubeCart version. I've seen it happen on two stores although it doesn't happen regularly. When building URL's if the site is accessed via IP address rather than domain name, the URL is built with the IP. This breaks the site if using SSL because the certificate is for the domain, not IP.

Is it possible to have the URL function swap out the IP for actual domain if an IP is to be returned?

**Comment by Noodleyman** [comment]
@abrookbanks , this needs to be re-opened. it still happens.

**Comment by abrookbanks** [comment]
Any steps to reproduce?

**Comment by Noodleyman** [comment]
It doesn't seem to happen as often as it used to, but for some reason I'm still seeing it happen in the wild. had it happen today on a prominent link on a busy site, which stops the link loading and browser throws a security error as the cert covers the domain, not IP.

**Comment by abrookbanks** [comment]
Yes I've seen it too. It's a rare beast.

**Comment by Noodleyman** [comment]
probably just needs a regex to stop IPv6 and IPv4 addresses in URL's being cached, unless an option is enabled in the store settings to keep existing behaviour. I seem to recall a specific customer who uses a IP rather than hostname.

**Comment by abrookbanks** [comment]
Yes I was thinking the same. :)

**Comment by havenswift-hosting** [comment]
I am also seeing this sometimes and on larger sites it obviously has a much bigger impact - any chance of trying to get a fix into 6.2.10 please ?

**Comment by abrookbanks** [comment]
I'm pretty confident this will finally resolve this one.
